### PR TITLE
Fix TypeScript types with Node16/NodeNext module resolution

### LINKED ISSUE #17964: Problem using from TypeScript as module
I cannot seem to import `@mdn/browser-compat-data` from TypeScript as a module.

`tsconfig.json`:
```json
{
	"compilerOptions": {
		"target": "ESNext",
		"module": "NodeNext",
		"moduleResolution": "NodeNext",
		"esModuleInterop": true,
		"resolveJsonModule": true,
		"allowSyntheticDefaultImports": true
	}
}
```
(Other ESM configurations also don't seem to work.)

`main.mts`:
```ts
import bcd from '@mdn/browser-compat-data' assert { type: 'json' };
import {CompatData} from '@mdn/browser-compat-data';
import * as bcdStar from '@mdn/browser-compat-data';
type CD = bcdStar.CompatData;
```

Compile: `tsc --pretty false`
Error:
```
main.mts(1,8): error TS1259: Module '"node_modules/@mdn/browser-compat-data/data"' can only be default-imported using the 'allowSyntheticDefaultImports' flag
main.mts(2,9): error TS2305: Module '"node_modules/@mdn/browser-compat-data/data"' has no exported member 'CompatData'.
main.mts(4,11): error TS2503: Cannot find namespace 'bcdStar'.
```

To me, this seems like it may also be a bug in TypeScript, but their module config documentation is not good, so I'm not sure. In this case I can also report it there, let me know.
EDIT: it says `"node_modules/@mdn/browser-compat-data/data"` in the error; apparently it looks at `data.json` due to `exports` in package.json and gets confused.

Maybe it would also be possible to change `index.ts` to `index.d.ts` where `./data.json` is not explicitly imported? If possible, that might also fix it, and it would also be very nice anyway, because `tsc` seems to spend a lot of time parsing the JSON file for no reason.

Version info:
- `@mdn/browser-compat-data` v5.2.5
- `typescript` v4.8.4

### PR DESCRIPTION
**Pull request by Gerrit0**
#### Summary

TypeScript with the Node16 or NodeNext setting for `"module"` in tsconfig.json will read the package exports map. This adds types entries to package.json so that it can be used in those settings. The `types` key is left in place so that the package types still work for users with `"module": "CommonJS"`.

TypeScript requires different declaration files (even though the same content works for both) so that it can distinguish between import/require. For clarity, the build script now writes to `require.d.ts` (not `.d.cts` for compatibility with old TS versions) and `import.d.mjs`

#### Test results and supporting details

Tested by adding the `build` directory as a dependency, and checking:

| Config | main | main types | forLegacyNode | forLegacyNode types |
| --- | --- | --- | --- | --- |
| `.mts` file with `NodeNext` | :white_check_mark: | :white_check_mark:  | :white_check_mark:  | :white_check_mark: |
| `.cjs` file with `NodeNext` | :white_check_mark: | :white_check_mark:  | :white_check_mark:  (require is error, can be dynamically imported) | :white_check_mark: (types can be imported) |
| `.ts` file with `CommonJS` using `import bcd from` transpiled into `require` | :white_check_mark: | :white_check_mark: | :x: (no support for `exports`, would have to rename files, anyone who runs into this should probably be using Node16/NodeNext anyways) | :x: |
| `.ts` file with `CommonJS` using `import x = require("@mdn/browser-compat-data")` | :x: (ts thinks the value is under .default, not many TS users even know this is a feature, don't think this matters) | :x: | :x: | :x: |

#### Related issues

Fixes #17964

### CODE REVIEW

File: scripts/release/build.ts [hunk 1]
```diff
@@ -174,7 +186,7 @@ export const createManifest = (): any => {
 const writeManifest = async () => {
   const dest = new URL('package.json', targetdir);
   const manifest = createManifest();
-  await fs.writeFile(dest, JSON.stringify(manifest));
+  await fs.writeFile(dest, JSON.stringify(manifest, null, 2));
```
Comment (Gerrit0) [inline scripts/release/build.ts:189]: The single line JSON was really annoying for debugging... and it's convenient to be able to actually read this file when checking on npmjs.org's "Code" tab. Not sure if it was intentionally minified, but even if so, writing with nice whitespace doesn't substantially change the package size.

Comment (queengooborg) [reply scripts/release/build.ts:189]: Could you provide a comparison of file size before and after this change?  Some of our consumers have very tight storage limitations they must adhere to, and it is a package used in many other projects, so it is important we keep the package as small as possible.

Comment (Gerrit0) [reply scripts/release/build.ts:189]: This is purely for the `package.json` file, this change has changed the package size less than the prior two releases which added a few entries to the `data.json` data. The main `data.json` file remains minified.

It is a difference of 190 characters, which is less than the size change from the recent commit adding `WebTransportSendStream`

Comment (queengooborg) [reply scripts/release/build.ts:189]: Ah, I misread the code and assumed this was altering `data.json`, my mistake.  `package.json` is fine to unminify.  👍

**Review by queengooborg — APPROVED** [review]
Alright, let's get this landed, thank you for your PR!  Welcome to BCD!

**Pull request merged**